src/main/flight/pid.c

```diff
@@ -987,8 +981,14 @@ void FAST_CODE pidController(const pidProfile_t *pidProfile, timeUs_t currentTim
                 pidRuntime.itermAccelerator = 0.0f; // no antigravity on yaw iTerm
             }
         }
-        const float iTermChange = (Ki + pidRuntime.itermAccelerator) * dynCi * pidRuntime.dT * itermErrorRate;
-        pidData[axis].I = constrainf(previousIterm + iTermChange, -pidRuntime.itermLimit, pidRuntime.itermLimit);
+
+        // axisCi implements input iTerm limiting based on pidSum compared to iTermLimit
+        float axisCi = (pidRuntime.itermLimit - fabsf(pidData[axis].Sum)) / pidRuntime.itermLimit;
+        // itermWindupPointInv sets the slope of the iterm attenuation.
+        axisCi = constrainf(axisCi * pidRuntime.itermWindupPointInv, 0.0f, 1.0f);
+
+        const float iTermChange = (Ki + pidRuntime.itermAccelerator) * axisCi * pidRuntime.dT * itermErrorRate;
```

Comment:
I don't like that  `axisCi` limits both I-Term increase and decrease.  If you somehow accumulate large enough I-term, it will never unwind (P will go against I term, so Sum should normally get into range where it is changing again. But can you proof that it will always be the case?)

using `axisCi` is increasing I-term direction is possible, but it will rectify noise (I term will accumulate negative spikes faster than positive ones). Will that alter flight behavior?

Reply:
You also assume that  `pidData[axis].I` can't accumulate more than  `pidRuntime.itermLimit`, but IMO that is not true - D-term may decrease `.Sum` (increasing axisCi), and I-term will still increase.

Reply:
Thanks, these are both good points.

Your first point I also am thinking about.  

With dynCi, the old method it is possible for motorMix to hit 100% and stop iTerm from changing.  iTerm limit by default in old code is 400.  With this limit, iTerm on one axis alone cannot saturate the motors.  But if two axes were to reach 400 on iTerm, then they can saturate the mixer, and then the gain of 0 will prevent iTerm from changing.

The modifications here share the same fundamental problem, that being:
If iTerm can grow to it's limit, it becomes very difficult for iTerm to go back below the limit, because the input gain factor is close to zero, and any change is slow.

For commanded moves, iTermRelax blocks iTerm change.  But in awkward landings or crashes or when stuck in a tree, we can get un-commanded ongoing error leading to iTerm accumulation to whatever limit exists at the time.  Mostly, D will become zero in the case of an ongoing un-correctible error of this kind, and FF will be zero, so pidSum is P + I alone.

The driver to get high iTerm is ongoing un-correctible error.  High error often means high P.  High P, especially ongoing high P, with the pidSum limitation, should reduce the limit for iTerm so it cannot grow to it's normal limit.  If the situation resolves and P falls, then the available limit should increase, providing room for iTerm to correct itself, or undo the accumulation.

As you point out, a workable solution is not trivial.  In normal flight, especially racers with batteries below turning very tight spirals, we can see significant iTerm buildup, and it is absolutely needed to control the move.  In fact most of pidSum then comes from iTerm.  Most freestyle / cruising doesn't need much sustained iTerm, but fast forward cruising in some frames has drag that requires iTerm to build up.  So we don't want to be dragging iTerm towards zero just because it is high.

I don't have a good solution for this dilemma, and it is not 'new'.

Your point about D being included in pidSum is noted.   For uncommanded issues, which is mostly where adverse iTerm accumulation is a more 'real' problem, P and D are in phase.  When the craft is 'stuck' both P and D will drop close to zero, and then iTerm will accumulate to its limit.

Maybe we can reset the iTerm value to the limit dynamically, so that if the value exceeds the limit, and the limit drops from new P or FF (new stick input), then we constrain iTerm itself to the new limit.  This could work.  Then anytime the user moves the sticks or changes the target angle, iTerm itself is dragged down.  I don't know.  I can try this idea.